server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -102,7 +106,49 @@ private static XContentBuilder mappings() {
         }
     }
 
-    public void putSynonymsSet(String resourceName, SynonymsSet synonymsset, ActionListener<PutSynonymsAction.Response> listener) {
+    public void getSynonymsSet(String resourceName, int from, int size, ActionListener<SynonymsSetResult> listener) {
+        client.prepareSearch(SYNONYMS_ALIAS_NAME)
+            .setQuery(QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName))
+            .setFrom(from)
+            .setSize(size)
+            .execute(listener.delegateFailure((searchResponseListener, searchResponse) -> {
+                final long totalSynonymRules = searchResponse.getHits().getTotalHits().value;
+                if (totalSynonymRules == 0) {
+                    listener.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
+                    return;
+                }
+                final SynonymRule[] synonymRules = Arrays.stream(searchResponse.getHits().getHits())
+                    .map(SynonymsManagementAPIService::hitToSynonymRule)
+                    .toArray(SynonymRule[]::new);
+                listener.onResponse(new SynonymsSetResult(totalSynonymRules, synonymRules));
+            }));
+    }
+
+    private static SynonymRule hitToSynonymRule(SearchHit hit) {
+        return new SynonymRule(
+            externalSynonymRuleId(hit.getId()),
+            (String) hit.getSourceAsMap().get(SynonymRule.SYNONYMS_FIELD.getPreferredName())
+        );
+    }
+
+    private static String externalSynonymRuleId(String internalId) {
```

Comment:
As Christoph mentioned, we need to internally use IDs that allow for different synonym sets to use the same synonym rule IDs without collision.

Synonym set id is used as a prefix to avoid collision. These two methods provide translation from the internal IDs to external IDs.

Reply:
+1